packages/lesswrong/components/review/ReviewVotingPage.tsx

```diff
@@ -278,49 +273,55 @@ const ReviewVotingPage = ({classes}: {
-    _id?: string|null,
-    postId: string,
-    change?: number,
-    set?: number
-  }) => {
-    const existingVote = _id ? dbVotes.find(vote => vote._id === _id) : null;
+  // TODO: This is untested in 2021
+  const dispatchQuadraticVote = async ({postId, change, set}: QuadraticVoteUpdate) => {
     await submitVote({
       variables: {postId, quadraticChange: change, newQuadraticScore: set, year: REVIEW_YEAR+"", dummy: false},
-      optimisticResponse: _id && {
-        __typename: "Mutation",
-        submitReviewVote: {
-          __typename: "ReviewVote",
-          ...existingVote,
-          quadraticScore: (typeof set !== 'undefined') ? set : ((existingVote?.quadraticScore || 0) + (change || 0))
-        }
-      }
     })
   }
 
   const { LWTooltip, Loading, ReviewVotingExpandedPost, ReviewVoteTableRow, SectionTitle, RecentComments, FrontpageReviewWidget } = Components
 
-  const [postOrder, setPostOrder] = useState<Map<number, number> | undefined>(undefined)
-  const reSortPosts = () => {
-    setPostOrder(new Map(getPostOrder(posts, useQuadratic ? quadraticVotes : votes, currentUser)))
+  const reSortPosts = useCallback(() => {
+    if (!postsResults) return
+    const randomPermutation = generatePermutation(postsResults.length, currentUser)
+    const newlySortedPosts = postsResults
+      .map((post, i) => ([post, randomPermutation[i]] as const))
+      .sort(([post1, permuted1], [post2, permuted2]) => {
+        const reviewedNotVoted1 = post1.reviewCount > 0 && !post1.currentUserReviewVote
+        const reviewedNotVoted2 = post2.reviewCount > 0 && !post2.currentUserReviewVote
+        if (reviewedNotVoted1 && !reviewedNotVoted2) return -1
+        if (!reviewedNotVoted1 && reviewedNotVoted2) return 1
+        if (post1.currentUserReviewVote < post2.currentUserReviewVote) return 1
+        if (post1.currentUserReviewVote > post2.currentUserReviewVote) return -1
+        if (permuted1 < permuted2) return -1;
+        if (permuted1 > permuted2) return 1;
+        return 0
+      })
+      .map(([post, _]) => post)
+    setSortedPosts(newlySortedPosts)
+    setPostsHaveBeenSorted(true)
+    
     captureEvent(undefined, {eventSubType: "postsResorted"})
-  }
-
-  // Re-sort in response to changes. (But we don't need to re-sort in response
-  // to everything exhaustively)
+  }, [currentUser, captureEvent, postsResults])
+  
+  const canInitialResort = !!postsResults
   useEffect(() => {
-    if (!!posts && useQuadratic ? !!quadraticVotes : !!votes) setPostOrder(new Map(getPostOrder(posts, useQuadratic ? quadraticVotes : votes, currentUser)))
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [!!posts, useQuadratic, !!quadraticVotes, !!votes])
+    reSortPosts()
+  }, [canInitialResort, reSortPosts])
+  
+  const quadraticVotes = useMemo(
+    () => sortedPosts?.map(post => ({
+      postId: post._id,
+      score: post.currentUserReviewVote,
+      type: 'QUADRATIC' as const
+    })),
+    [sortedPosts]
+  )
```

Comment:
I don't know much about quadratic voting, but is it ok for all posts to be given a "vote", even the ones the user didn't actually vote on it? Does that impact the renormalization?

Reply:
Good catch! Unfortunately, quadratic voting is hard to test because it's awaiting a decision about whether we should even use it and isn't known to work currently. However, I think this is probably another case where I should check for non-null currentUserReviewVote, so thanks!